### PR TITLE
Render posts before documents, remove threading

### LINKED ISSUE #138: Get the content of a post
It would be nice to be able to get the individual contents of each post for some kind of main page feature. So like:

```liquid
{% for post in posts %}
<article>
  <h2>{{post.title}}</h2>
  {{post.content}}
</article>
{% endfor %}
```

### PR DESCRIPTION
**Pull request by johannhof**
In #138 and #145 I noticed that it might be beneficial to render posts
before documents, because documents might want to access post
information such as content or excerpt that requires all posts to be
rendered.

This also removes the admittedly naive approach of spawning a thread per
document which at least by my measurements adds way too much overhead to
be of any use. I will open a new issue for experimenting with ways to
add good parallelism.

Fixes #138 by adding a "content" attribute to posts.

### CODE REVIEW

**Comment by johannhof** [comment]
@nott Would you like review this or tell me what you think? :)

If you check out the way I assign a new "content" attribute this could be how we could also add the excerpt in your PR without having to re-render the markdown.

It's still missing tests for the "content" part, I'll add those tomorrow.

File: src/cobalt.rs [hunk 1]
```diff
@@ -115,44 +113,47 @@ pub fn build(config: &Config) -> Result<()> {
...
+
     // sort documents by date, if there's no date (none was provided or it couldn't be read) then
     // fall back to the default date
-    documents.sort_by(|a, b| b.date.unwrap_or(default_date).cmp(&a.date.unwrap_or(default_date)));
+    posts.sort_by(|a, b| b.date.unwrap_or(default_date).cmp(&a.date.unwrap_or(default_date)));
 
     // check if we should create an RSS file and create it!
     if let &Some(ref path) = &config.rss {
-        try!(create_rss(path, dest, &config, &documents));
+        try!(create_rss(path, dest, &config, &posts));
```
Comment (nott) [inline src/cobalt.rs]: i think it makes sense to `create_rss` after we insert content, so we can add content (or excerpt) to the feed

**Comment by nott** [comment]
i did some naive performance testing and this pr actually improved performance on my very small blog

File: src/cobalt.rs [hunk 2]
```diff
@@ -115,44 +113,47 @@ pub fn build(config: &Config) -> Result<()> {
...
     }
 
     // these are the attributes of all documents that are posts, so that they can be
     // passed to the renderer
-    // TODO: do we have to clone these?
-    let post_data: Vec<Value> = documents.iter()
-        .filter(|x| x.is_post)
+    let mut post_data: Vec<Value> = posts.iter()
```
Comment (nott) [inline src/cobalt.rs]: I understand the naming comes from previous revisions... however

name `post_data` is a little bit confusing. i'd call this at least `posts_data`

File: src/cobalt.rs [hunk 3]
```diff
@@ -115,44 +113,47 @@ pub fn build(config: &Config) -> Result<()> {
...
 
-    // thread handles to join later
-    let mut handles = vec![];
+    let simple_post_data = post_data.clone();
 
-    // generate documents (in parallel)
-    crossbeam::scope(|scope| {
-        let post_data = Arc::new(post_data);
+    trace!("Generating posts");
+    for (post, data) in posts.iter().zip(post_data.iter_mut()) {
```
Comment (nott) [inline src/cobalt.rs]: we have 3 copies of posts attributes at this point (post_data, simple_post_data, posts)

what if we

* use 2 copies: `simple_post_data` and posts
* `post.attributes.insert` `content` in the posts loop
* and create `post_data` vector once again after the loop

imho that will simplify the code - we won't need the `unreachable!()` below

**Comment by nott** [comment]
great pr

your scepsis about #145 was fair

File: src/cobalt.rs (hunk 1 above)
Comment (nott) [inline src/cobalt.rs]: however this should def be a separate pr

File: src/cobalt.rs (hunk 3 above)
Comment (johannhof) [inline src/cobalt.rs]: **Awesome idea**, I think we can even improve on top of this and get down to cloning posts only once!

File: src/cobalt.rs (hunk 2 above)
Comment (johannhof) [inline src/cobalt.rs]: Good point

File: src/cobalt.rs (hunk 1 above)
Comment (johannhof) [inline src/cobalt.rs]: Yes, I agree. Happy to take a PR if you want to take care of it :)

Comment (nott) [inline src/cobalt.rs]: :+1:

**Comment by johannhof** [comment]
@nott Thanks for the great comments. I think we ended up with faster and much more readable and code.

> i did some naive performance testing and this pr actually improved performance on my very small blog

Yeah I implemented the threading at a time when I was just blindly parallelizing my Rust code because it's so easy. Turns out oftentimes the overhead (performance + cognitive) is larger than if you just do it sequentially.

**Comment by nott** [comment]
LGTM

**Comment by nott** [comment]
@johannhof

i've just realized there is a bug here: `content` contains html taken from layouts
that won't be useful for the case mentioned in #138
